scripts/keys.coffee

```diff
@@ -10,95 +10,247 @@
+				if key_holder is name
+					key_holder = ""
+				robot.brain.set("keyBlue",key_holder)	
+			else if keyname is 'yellow'
+				key_holder = keyYellow()
+				if key_holder is name
+					key_holder = ""	
+				robot.brain.set("keyYellow",key_holder)
+			else if keyname is 'green'
+				key_holder = keyGreen()
+				if key_holder is name
+					key_holder = ""	
+				robot.brain.set("keyGreen",key_holder)
+			else if keyname is 'orange'
+				key_holder = keyOrange()
+				if key_holder is name
+					key_holder = ""	
+				robot.brain.set("keyOrange",key_holder)
+			if typeof user is 'object'
+				if key_holder is ""
+					msg.send "Okay #{name} doesn't have #{keyname} keys. Who got the #{keyname} keys then?"
+				else
+					msg.send "Yes , I know buddy, its because #{key_holder} has got the #{keyname} keys"
 
-	robot.respond /(.+) (has|have) (the key|key|keys|a key)/i , (msg)->
+    #3this section belongs to "$name has keys"
+	robot.respond /(.+) (has the|have the|has|have) (.+) (key|keys)/i , (msg)->
 		othername = msg.match[1]
+		keyname = msg.match[3]
 		name = msg.message.user.name
-		k = key()
 		unless othername in ["who", "who all","Who", "Who all" , "i" , "I" , "i don't" , "i dont" , "i do not" , "I don't" , "I dont" , "I do not"]
-			if othername is 'you'
-				msg.send "How am I supposed to take those keys? #{name} is a liar!"
-			else if othername is robot.name
-				msg.send "How am I supposed to take those keys? #{name} is a liar!"
-			else
-				users = robot.brain.usersForFuzzyName othername
-				if users.length is 1
-					robot.brain.set "key", users[0].name
-					msg.send "Okay, so now the keys are with #{users[0].name}"
-				else if users.length > 1
-					msg.send getAmbiguousUserText users
+			unless keyname not in ["red", "blue", "yellow", "green", "orange"]
+				users = robot.brain.userForName othername
+				if users is null
+					key_holder = null
 				else
-					msg.send "#{othername}? Never heard of 'em"
-
-	robot.respond /(i|I) (have given|gave|had given) (the key|key|keys|a key|the keys) to (.+)/i , (msg)->
-		othername = msg.match[4]
+					key_holder = users.name
+				if key_holder is null
+					if othername is 'you'
+						msg.send "How am I supposed to take those #{keyname} keys? #{name} is a liar!"
+					else if othername is robot.name
+						msg.send "That's utter lies! How can you blame a bot to have the keys?"
+					else
+						msg.send "I don't know anyone by the name #{othername}"
+				else
+					if keyname is 'red'
+						robot.brain.set("keyRed",key_holder)
+					else if keyname is 'blue'
+						robot.brain.set("keyBlue",key_holder)
+					else if keyname is 'yellow'
+						robot.brain.set("keyYellow",key_holder) 
+					else if keyname is 'green'
+						robot.brain.set("keyGreen",key_holder) 
+					else if keyname is 'orange'
+						robot.brain.set("keyOrange",key_holder) 
+					msg.send "Okay, so now the #{keyname} keys are with #{othername}"	
+    #4this section belongs to "i gave the keys to $name"
```

Comment:
Also implement: "I gave keys to {key_holder}"
Basically, you will automatically check the colour and assign that to the {key_holder}

Reply:
ok